### DISCUSSION TITLE
Error: connect ECONNREFUSED

### DISCUSSION THREAD

**Discussion by ether-wan** [apocas/dockerode#725] [Q&A] [answered]
Hello, i'm trying to launch container w/ dockerode on Mac OS

```
const docker = new Docker({host: 'http://0.0.0.0', port: 3000, socketPath: '/var/run/docker.sock'});

    const container = await docker.createContainer({
        Image: 'app-bot',
        HostConfig: {
            PortBindings: {
                [`${3000}/http`]: [{ HostPort: 50000 }]
            }
        }
    });
    await container.start();
```

I'm getting an error : 

`
Error: connect ECONNREFUSED 0.0.0.0:3000
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1247:16) {
  errno: -61,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '0.0.0.0',
  port: 3000
}
`

The docker service run with colima ( https://github.com/abiosoft/colima ) 

The deployment of the image works with  `docker run -p 50000:3000 app-bot` command on the terminal

Thanks !!

**Comment by danisharora099** [comment]
you also need to use `ExposedPorts` along with `PortBindings`

**Comment by ether-wan** [comment]
Thanks @danisharora099 for your reply. I'm not sure where I should use `ExposedPorts`, can you give me an example please

**Comment by danisharora099** [comment]
> Thanks @danisharora099 for your reply. I'm not sure where I should use `ExposedPorts`, can you give me an example please


```
const container = await docker.createContainer({
        Image: "app-bot",
        HostConfig: {
          PortBindings: {
            [`${3000}/http`]: [{ HostPort: 50000 }],
          },
        },
        ExposedPorts: {
          [`${3000}/http`]: {},
        },
      });
    await container.start();
```


Also, not sure if `/http` is a supported protocol. You could probably get away with subsituting `http` with `tcp`

**Comment by ether-wan** [comment]
> 
it doesn't work for me, I still get the ECONNREFUSED error. Are there any specific permissions to give ? 

```
Error: connect ECONNREFUSED 0.0.0.0:3000
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1247:16) {
  errno: -61,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '0.0.0.0',
  port: 3000
}
```

**Comment by danisharora099** [comment]
Did you try it with `/tcp`? @reptincel-cpu

**Comment by ether-wan** [comment]
> Did you try it with `/tcp`? @reptincel-cpu

Yes ! I get the same error ...

**Comment by aelesbao** [comment, marked as answer]
@reptincel-cpu I had the same issue, but I discovered that Colima installs itself as a new `docker context`. That's the output of my `docker context ls` command:

```
NAME       DESCRIPTION                               DOCKER ENDPOINT                                      ERROR
colima *   colima                                    unix://$HOME/.colima/default/docker.sock
default    Current DOCKER_HOST based configuration   unix:///var/run/docker.sock
```

As you can see, the socket path is different from the one used by Docker. You can either change the `socketPath` param when instantiating a new `Docker` class, or you can create a symlink to Colima's socket:

```sh
sudo ln -sf /var/run/docker.sock ~/.colima/default/docker.sock
```
